pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -132,6 +147,153 @@ func storageMatches(set *apps.StatefulSet, pod *v1.Pod) bool {
 	return true
 }
 
+// claimOwnerMatchesSetAndPod returns false if the ownerRefs of the claim are not set consistently with the
+// PVC deletion policy for the StatefulSet.
+func claimOwnerMatchesSetAndPod(claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) bool {
+	switch set.Spec.PersistentVolumeClaimDeletePolicy {
+	case apps.Retain:
+		if hasOwnerRef(claim, set) ||
+			hasOwnerRef(claim, pod) {
+			return false
+		}
+	case apps.DeleteOnStatefulSetDeletionOnly:
+		if !hasOwnerRef(claim, set) ||
+			hasOwnerRef(claim, pod) {
+			return false
+		}
+	case apps.DeleteOnScaledownOnly:
+		if hasOwnerRef(claim, set) {
+			return false
+		}
+		podScaledDown := getOrdinal(pod) >= int(*set.Spec.Replicas)
+		if podScaledDown != hasOwnerRef(claim, pod) {
+			return false
+		}
+	case apps.DeleteOnScaledownAndStatefulSetDeletion:
+		podScaledDown := getOrdinal(pod) >= int(*set.Spec.Replicas)
+		// If a pod is scaled down, there should be no set ref and a pod ref;
+		// if the pod is not scaled down it's the other way around.
+		if podScaledDown == hasOwnerRef(claim, set) {
+			return false
+		}
+		if podScaledDown != hasOwnerRef(claim, pod) {
+			return false
+		}
+	default:
+		klog.Errorf("Unknown policy %v!", set.Spec.PersistentVolumeClaimDeletePolicy)
+	}
+	return true
+}
+
+// updateClaimOwnerRefForSetAndPod updates the ownerRefs for the claim according to the deletion policy of
+// the StatefulSet. Returns true if the claim was changed and should be updated and false otherwise.
+func updateClaimOwnerRefForSetAndPod(claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) bool {
+	needsUpdate := false
+	// Sometimes the version and kind are not set pod.TypeMeta. These are necessary for the ownerRef.
+	// TODO: there must be a better way to do this other than hardcoding the pod version?
+	podMeta := pod.TypeMeta
+	if podMeta.APIVersion == "" {
```

Comment:
Does this happen in a real cluster or only unit tests?

Reply:
Yes, this was happening in a real cluster. Maybe this means something has to wait a reconcile loop for the pod to be created?